Add mutual TLS docs (#1132)
elastic/ingest-docs: docs/en/ingest-management/security/mutual-tls.asciidoc

[[mutual-tls]]
= {agent} deployment models with mutual TLS

> Mutual Transport Layer Security (TLS) provides a higher level of security and trust compared to one-way TLS, where only the server presents a certificate. It ensures that not only the server is who it claims to be, but the client is also authenticated. This is particularly valuable in scenarios where both parties need to establish trust and validate each other's identities, such as in secure API communication, web services, or remote authentication.

Author: Changed to `mTLS`.

* <<mutual-tls-overview>>
* <<mutual-tls-on-premise>>
* <<mutual-tls-cloud>>
* <<mutual-tls-cloud-proxy>>
* <<mutual-tls-on-premise-hosted-es>>

//[source,shell]
//----
//example
//----

//image::images/fleet-server-certs.png[Screen capture of a folder called fleet-server that contains two files: fleet-server.crt and fleet-server.key]

[discrete]
[[mutual-tls-overview]]
== Overview

With mutual TLS the following authentication and certification verification occurs:

* **Client Authentication**: The client presents its digital certificate to the server during the TLS handshake. This certificate is issued by a trusted Certificate Authority (CA) and contains the client's public key.
* **Server Authentication**: The server also presents its digital certificate to the client, proving its identity and sharing its public key. The server's certificate is also issued by a trusted CA.

> * **Certificate Verification**: Both the client and server verify each other's certificates by checking the digital signatures against the CA's public key. This process confirms the authenticity of the certificates and, by extension, the identities of the client and server.

Author: I think it's probably worth noting that the CAs need not be the same, so I've add that as a note.

> Both the client and server verify each other's certificates by checking the digital signatures against the CAs' public key (note that the client and server need not use the same CA).

> * **Certificate Verification**: Both the client and server verify each other's certificates by checking the digital signatures against the CA's public key. This process confirms the authenticity of the certificates and, by extension, the identities of the client and server.

Author: I've removed that sentence.

{fleet}-managed {agent} has two main connections to ensure correct operations:

* Connectivity to {fleet-server} (the control plane, to check in, download policies, and similar).
* Connectivity to an Output (the data plane, such as {es} or {ls}).

In order to bootstrap, {agent} initially must establish a secure connection to the {fleet-server}, which can reside on-premises or in {ecloud}. This connectivity verification process ensures the agent's authenticity. Once verified, the agent receives the policy configuration. This policy download equips the agent with the knowledge of the other components it needs to engage with. For instance, it gains insights into the output destinations it should write data to.

> If mutual TLS (mTLS) is a requirement, {agent} must first establish an mTLS connection with {fleet-server}, with both client and server exchanging certificates and validating one another. Once the policy configuration is in place, it possesses the necessary details to establish an mTLS connection with the specific output it's configured to use. In the case of {fleet}-managed {agents}, certificates and certificate authorities essential for client-server authentication are configured through the {fleet} application in the {kib} user interface. As previously mentioned, the initial step involves establishing connectivity between {agent} and the {fleet-server}, allowing the subsequent configuration to take effect.

Author: Sounds good! I've removed a couple of paragraphs and added the precedence:

![Screenshot 2024-06-19 at 3 35 10 PM](https://github.com/elastic/ingest-docs/assets/41695641/e5f1b173-f4e9-47d6-b2c0-84d6240841bd)

[discrete]
[[mutual-tls-on-premise]]
== On-premise deployments

image::images/mutual-tls-on-prem.png[Diagram of mutual TLS on premise deployment model]

Refer to the steps in <<secure-connections,Configure SSL/TLS for self-managed {fleet-server}s>>. To configure mutual TLS, include the following additional parameters when you install {agent} and {fleet-server}.

[discrete]
=== {agent} settings
During {agent} installation on premise use the following options:

[cols="1,1"]
|===
|`-certificate-authorities`
|List of CA certificates that are trusted when {fleet-server} connects to {agent} 

|`-elastic-agent-cert`
|{agent} certificate to present to {fleet-server} during authentication

|`-elastic-agent-cert-key`
|{agent} certificate key to present to {fleet-server}
|===

[discrete]
=== {fleet-server} settings
During {fleet-server} installation on-premise {fleet-server} authenticates with {es} and {agents}. You can use the following CLI options to facilitate these secure connections:

[cols="1,1"]
|===
|`-fleet-server-es-ca`
|CA to use for the {es} connection 

|`-fleet-server-es-cert`
|{fleet-server} certificate to present to {es}

> ∫

Author: Funny - I wouldn't know how to type that character if I wanted to. I can only do it by accident. :-)

|`-fleet-server-es-cert-key`
|{fleet-server} certificate key to present to {es}

> |`-certificate-authorities`

Author: I've changed this to:

> -certificate-authorities
> List of CA certificates that are trusted when {agent} connects to {fleet-server} and when {fleet-server} validates the {agent} identity.

|List of CA certificates that are trusted when {agent} connects to {fleet-server}

|`-fleet-server-cert`
|{fleet-server} certificate to present to {agents} during authentication

|`-fleet-server-cert-key`

> |{fleet-server} certificate key to present to {agents}  

Author: Changed to:

> |-fleet-server-cert-key
> {fleet-server's} private certificate key used to decrypt the certificate

I also made this change where the setting appears further down  in the "on-prem hosted ES" section.

|===

[discrete]
=== {fleet} settings:

In {kib}, navigate to {fleet}, open the **Settings** tab, and choose the **Output** that you'd like to configure.
In the **Advanced YAML configuration**, add the following settings:

[cols="1,1"]
|===

> |`-certificate-authorities`

Reviewer: [Blocker]
the correct key is:
```suggestion
|`ssl.certificate-authorities`
```

Author: I've changed all instances of `-certificate-authorities` to `ssl.certificate-authorities`.

|List of CA certificates that are trusted when {fleet-server} connects to {agent}

|`ssl.certificate`
|This certificate will be passed down to all the agents that have this output configured in their policy. This certificate is used by the agent when establishing mTLS to the output. 

You may either apply the full certificate, in which case all the agents get the same certificate OR alternatively point to a local directory on the agent where the certificate resides, if the certificates are to be unique per agent.

|`ssl.key`

> |This certificate key will be passed down to all the agents that have this output configured in their policy. This key is used by the agent when establishing mTLS to the output. 

Author: Changed to:

> |`ssl.key`
> |This certificate key will be passed down to all the agents that have this output configured in their policy. The certificate key is used to decrypt the SSL certificate.

> Similar to the certificate, the certificate key may be stored in a local directory.

> |This certificate key will be passed down to all the agents that have this output configured in their policy. This key is used by the agent when establishing mTLS to the output. 

Author: I've put this in as a note, below the settings table:

<img width="692" alt="Screenshot 2024-06-19 at 4 14 44 PM" src="https://github.com/elastic/ingest-docs/assets/41695641/7ef63948-7378-45cf-8d49-215773fda308">

Similarly the key may be stored in a local directory.
|===

image::images/mutual-tls-onprem-advanced-yaml.png[Screen capture of output advanced yaml settings]

[discrete]
[[mutual-tls-cloud]]
== {fleet-server} on {ecloud}

> In this deployment model, all traffic ingress into {ecloud} has its TLS connection terminated at an Elastic-managed Cloud Proxy. Cloud Proxy is not provisioned on a per tenant basis so at this point a client specific certificate can NOT be used in this proxy.

Author: Added exactly as suggested.

image::images/mutual-tls-cloud.png[Diagram of mutual TLS on cloud deployment model]

We currently don't support mTLS in this deployment model. An alternate deployment model is shown below where you can deploy your own secure proxy where TLS connections are terminated.

[discrete]
[[mutual-tls-cloud-proxy]]

> == {fleet-server} on {ecloud} using a proxy

Author: I didn't change anything, as I think we use the same titles in other spots about the various deployment models, but this is really good to know!

> In this scenario, unlike the one above, you have access to the proxy and therefore can configure it to do a TLS termination.

Author: Fixed exactly as suggested.

image::images/mutual-tls-cloud-proxy.png[Diagram of mutual TLS on cloud deployment model with a proxy]

[discrete]
=== {agent} settings
During {agent} installation on premise use the following options:

[cols="1,1"]
|===
|`-certificate-authorities`

> |List of CA certificates that are trusted when {fleet-server} connects to {agent} 

Author: Changed to:

> ssl.certificate-authorities
> List of CA certificates that are trusted when {agent} connects to {fleet-server} or to the proxy between {agent} and {fleet-server}.

|`-elastic-agent-cert`
|{agent} certificate to present to {fleet-server} during authentication

|`-elastic-agent-cert-key`

> |{agent} certificate key to present to {fleet-server}

Author: Changed to:

> -elastic-agent-cert
> {agent} certificate to present  during authentication to {fleet-server} or to the proxy between {agent} and {fleet-server}

> -elastic-agent-cert-key
> {agent}'s private certificate key used to decrypt the certificate

|===

[discrete]
[[mutual-tls-on-premise-hosted-es]]
== {fleet-server} on-premise and Hosted Elasticsearch Service

> In some scenarios you may want to deploy {fleet-server} on your own premises. In this case, you're able to provide your own certificates and certificate authority to facilitate mTLS between {fleet-server} and {agent}.

Author: Good point! I changed this to use `enable`:

> In some scenarios you may want to deploy {fleet-server} on your own premises. In this case, you're able to provide your own certificates and certificate authority to enable mTLS between {fleet-server} and {agent}.

> However, as with the <<mutual-tls-cloud,{fleet-server} on {ecloud}>> use case, the data plane connections are terminated at Elastic's Cloud Proxy, which is not a multi-tenanted service and therefore can't provide per-user certificates.  

Author: I changed this to:

> However, as with the <<mutual-tls-cloud,{fleet-server} on {ecloud}>> use case, the data plane connections are has its TLS connection terminated at the {ecloud} boundary. {ecloud} is not a multi-tenanted service and therefore can't provide per-user certificates.

image::images/mutual-tls-fs-onprem.png[Diagram of mutual TLS with Fleet Server on premise and hosted Elasticsearch Service deployment model]

Similar to the {fleet-server} on {ecloud} use case, a secure proxy can be placed in such an environment to terminate the TLS connections and satisfy the mTLS requirements.

image::images/mutual-tls-fs-onprem-proxy.png[Diagram of mutual TLS with Fleet Server on premise and hosted Elasticsearch Service deployment model with a proxy]

[discrete]
=== {agent} settings
During {agent} installation on premise use the following options, similar to <<mutual-tls-on-premise,{agent} deployment on premises>>:

[cols="1,1"]
|===
|`-certificate-authorities`

> |List of CA certificates that are trusted for when {fleet-server} connects to {agent} 

Author: Changed to:

> ssl.certificate-authorities
> List of CA certificates that are trusted for when {agent} connects to {fleet-server}

|`-elastic-agent-cert`
|{agent} certificate to present to {fleet-server} during authentication

|`-elastic-agent-cert-key`

> |{agent} certificate key to present to {fleet-server}

Author: Changed to:

> -elastic-agent-cert-key
> {agent}'s private certificate key used to decrypt the certificate

|===

[discrete]
=== {fleet-server} settings
During {fleet-server} installation on-premise use the following options so that {fleet-server} can authenticate itself to the agent and then also to the secure proxy server:

[cols="1,1"]
|===
|`-fleet-server-es-ca`
|CA to use for the {es} connection, via secure proxy. This CA is used to authenticate the TLS connection from a secure proxy

|`-certificate-authorities`
|List of CA certificates that are trusted when {agent} connects to {fleet-server}

|`-fleet-server-cert`
|{fleet-server} certificate to present to {agents} during authentication

|`-fleet-server-cert-key`

> |{fleet-server} certificate key to present to {agents}  

Author: Changed to:

> -fleet-server-cert-key
> {fleet-server}'s private certificate key used to decrypt the certificate

|===

[discrete]
=== {fleet} settings

> This is the same as what's described for <<mutual-tls-on-premise,on premise deployments>>. The main difference is that you need to use certificates that are accepted by the secure proxy, as that is where the TLS connection will be terminated.

Reviewer: [Suggestion]
Personally I'd rather focus more on the mTLS than the TLS termination itself. Also because you can have TLS termination without mTLS. Something in the line of:

```suggestion
This is the same as what's described for <<mutual-tls-on-premise,on premise deployments>>. The main difference is that you need to use certificates that are accepted by the secure proxy, as the mTLS is setup between the agent and the secure proxy.
```

Author: Changed to:

> This is the same as what's described for <<mutual-tls-on-premise,on premise deployments>>. The main difference is that you need to use certificates that are accepted by the secure proxy, as the mTLS is set up between the agent and the secure proxy.

//To facilitate the bootstrapping process and enable {agent} to establish an mTLS connection with {fleet-server}, all certificates and certificate authorities are configured using command-line parameters during the agent installation. Once the mTLS connection between {agent} and the {fleet-server} is established, the policy configuration enables the establishment of the mTLS connection between {agent} and the designated output as well.

When mTLS is required, the secure setup between {agent}, {fleet}, and {fleet-server} is configured through the following steps:

. mTLS is enabled.
. The initial mTLS connection between {agent} and {fleet-server} is configured when {agent} is enrolled, using the parameters passed through the `elastic-agent install` or `elastic-agent enroll` command.

> . Once enrollment is completed, {agent} downloads the initial agent policy from {fleet-server}. That policy contains the mTLS configuration settings that will take precedence over those used during enrollment: This includes both the mTLS settings used for connectivity between {agent} and {fleet-server} (and the {fleet} application in {kib}, for {fleet}-managed (agent)), and the settings used between {agent} and it's specified output.

Author: Definitely worth mentioning! I've re-organized it a little bit:

![Screenshot 2024-06-20 at 11 50 31 AM](https://github.com/elastic/ingest-docs/assets/41695641/a14628ea-5840-40e1-932b-8193d6cdb3b7)

[discrete]
[[mutual-tls-on-premise]]
== On-premise deployments

image::images/mutual-tls-on-prem.png[Diagram of mutual TLS on premise deployment model]

Refer to the steps in <<secure-connections,Configure SSL/TLS for self-managed {fleet-server}s>>. To configure mutual TLS, include the following additional parameters when you install {agent} and {fleet-server}.

[discrete]
=== {agent} settings
During {agent} installation on premise use the following options:

[cols="1,1"]
|===

> |`ssl.certificate-authorities`

Reviewer: ```suggestion
|`--certificate-authorities`
```

Author: Fixed all instances.

|List of CA certificates that are trusted when {fleet-server} connects to {agent} 

> |`-elastic-agent-cert`

Reviewer: ```suggestion
|`--elastic-agent-cert`
```

Author: Fixed all instances.

|{agent} certificate to present to {fleet-server} during authentication

> |`-elastic-agent-cert-key`

Reviewer: ```suggestion
|`--elastic-agent-cert-key`
```

Author: Fixed all instances.

[discrete]
[[mutual-tls-cloud-proxy]]
== {fleet-server} on {ecloud} using a proxy

In this scenario, where you have access to the proxy, you can configure mTLS between the agent and your proxy.

image::images/mutual-tls-cloud-proxy.png[Diagram of mutual TLS on cloud deployment model with a proxy]

[discrete]
=== {agent} settings
During {agent} installation on premise use the following options:

[cols="1,1"]
|===
|`ssl.certificate-authorities`
|List of CA certificates that are trusted when {agent} connects to {fleet-server} or to the proxy between {agent} and {fleet-server}

|`-elastic-agent-cert`
|{agent} certificate to present  during authentication to {fleet-server} or to the proxy between {agent} and {fleet-server}

|`-elastic-agent-cert-key`
|{agent}'s private certificate key used to decrypt the certificate
|===

[discrete]
[[mutual-tls-on-premise-hosted-es]]
== {fleet-server} on-premise and Hosted Elasticsearch Service

In some scenarios you may want to deploy {fleet-server} on your own premises. In this case, you're able to provide your own certificates and certificate authority to enable mTLS between {fleet-server} and {agent}.

> However, as with the <<mutual-tls-cloud,{fleet-server} on {ecloud}>> use case, the data plane connections are has its TLS connection terminated at the {ecloud} boundary. {ecloud} is not a multi-tenanted service and therefore can't provide per-user certificates.  

Author: I think "connections have their TLS connection" is also a bit confusing, so I've rewritten to:

![Screenshot 2024-06-20 at 11 53 39 AM](https://github.com/elastic/ingest-docs/assets/41695641/717101e1-7930-469e-a2be-744c1716dd31)
